chronos/asyncproc.nim

```diff
@@ -0,0 +1,1323 @@
+          return i
+      len(a)
+
+    while true:
+      if osdefs.getcwd(cstring(res), bufsize) != nil:
+        setLen(res, strLength(res))
+        return ok(res)
+      else:
+        let errorCode = osLastError()
+        if errorCode == osdefs.ERANGE:
+          bufsize = bufsize shl 1
+          doAssert(bufsize >= 0)
+          res = newString(bufsize)
+        else:
+          return err(errorCode)
+
+  proc setCurrentDirectory(dir: string): AsyncProcessResult[void] =
+    let res = osdefs.chdir(cstring(dir))
+    if res == -1:
+      return err(osLastError())
+    ok()
+
+  proc closeThreadAndProcessHandle(p: AsyncProcessRef
+                                  ): AsyncProcessResult[void] =
+    discard
+
+  proc startProcess*(command: string, workingDir: string = "",
+                     arguments: seq[string] = @[],
+                     environment: StringTableRef = nil,
+                     options: set[AsyncProcessOption] = {},
+                     stdinHandle = ProcessStreamHandle(),
+                     stdoutHandle = ProcessStreamHandle(),
+                     stderrHandle = ProcessStreamHandle(),
+                    ): Future[AsyncProcessRef] {.async.} =
+    var
+      pid: Pid
+      pipes = preparePipes(options, stdinHandle, stdoutHandle,
+                           stderrHandle).valueOr:
+        raiseAsyncProcessError("Unable to initialze process pipes",
+                               error)
+      sa = pipes.initSpawn(options).valueOr:
+        discard closeProcessHandles(pipes, options, OSErrorCode(0))
+        await pipes.closeProcessStreams(options)
+        raiseAsyncProcessError("Unable to initalize spawn attributes", 0)
+
+    let
+      (commandLine, commandArguments) =
+        if AsyncProcessOption.EvalCommand in options:
+          let args = @[chronosProcShell, "-c", command]
+          (chronosProcShell, allocCStringArray(args))
+        else:
+          var res = @[command]
+          for arg in arguments.items():
+            res.add(arg)
+          (command, allocCStringArray(res))
+      commandEnv =
+        if isNil(environment):
+          envToCStringArray()
+        else:
+          envToCStringArray(environment)
+
+    var currentError: OSErrorCode
+    var currentDir: string
+
+    try:
+      currentDir =
+        if len(workingDir) > 0:
+          # Save current working directory and change it to `workingDir`.
+          let cres = getCurrentDirectory()
+          if cres.isErr():
+            raiseAsyncProcessError("Unable to obtain current directory",
+                                   cres.error())
+          let sres = setCurrentDirectory(workingDir)
+          if sres.isErr():
+            raiseAsyncProcessError("Unable to change current directory",
+                                   sres.error())
+          cres.get()
+        else:
+          ""
+
+      if AsyncProcessOption.EchoCommand in options:
+        echo getFullCommand(commandLine, commandArguments)
```

Comment:
This feature is quite unusual.
If the user needs it, it would be very easy to implement it as a wrapper of `startProcess`.

Reply:
Fixed.